### PR TITLE
[XB1] Fix skipped OnDecoderDrained task.

### PR DESCRIPTION
**Pull request by victorpasoshnikov**
b/284359403

### CODE REVIEW

**Comment by victorpasoshnikov** [comment]
The following has place in sw (vpx, av1) decoders.
In GpuVideoDecoderBase::Reset() we have call consequence
1)    decoder_thread_->job_queue()->Schedule(
        std::bind(&GpuVideoDecoderBase::DrainDecoder, this)); <- put task DrainDecoder into thread's job queue
 2)   decoder_thread_.reset();   <- Put task JobQueue::StopSoon and join thread
But  DrainDecoder causes OnDecoderDrained() in different thread (underlaying decoder's output thread) and the task OnDecoderDrained() can be created AFTER JobQueue::StopSoon task and can't be processed.
To avoid this situation I had to add extra sync objects to make sure that decoder_thread_.reset() will be called after OnDecoderDrained() call

**Review by TyHolc — APPROVED** [review]
Looks good to me, I'll leave it to the media team to give a final review.

File: starboard/xb1/shared/gpu_base_video_decoder.cc [hunk 1]
```diff
@@ -475,6 +476,16 @@ void GpuVideoDecoderBase::Reset() {
     decoder_status_cb_(kReleaseAllFrames, nullptr);
     decoder_thread_->job_queue()->Schedule(
         std::bind(&GpuVideoDecoderBase::DrainDecoder, this));
+    {
+      // Before calling decoder_thread_.reset() wait until
+      // OnDecoderDrained is put into job queue.
+      ScopedLock lock(drained_mutex_);
+      if (!drained_condition_.WaitTimed(3 * kSbTimeSecond)) {
```
[the remark this answers is not in the document]
Comment (victorpasoshnikov) [reply starboard/xb1/shared/gpu_base_video_decoder.cc:483]: I suppose Reset() call means that all process in decoder finished and decoder's condition become neutral. I think that calling drain() is universal method to ensure that all tasks in decoder are completed. I can suggest more simple solution to resolve current issue.
It is a fact that the DrainDecoderInternal() is a sync command. After this command we can be sure that the drain is really done. So we don't need to wait zero image in Av1(vpx)VideoDecoder::OnFrameReady. Zero image goes always right before  DecodeInternal(nullptr) returns.
I will offer the concrete solution in the next commit in this PR & new CL in Gerrit

**Comment by victorpasoshnikov** [comment]
This PR must be merged together with https://lbshell-internal-review.googlesource.com/c/cobalt_src/+/267860

File: starboard/xb1/shared/gpu_base_video_decoder.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply starboard/xb1/shared/gpu_base_video_decoder.cc]: Ok, the new solution is simple enough. Just a last double check, can we confirm the drain operation are all finished synchronous in DecodeInternal() for VpxVideoDecoder, Av1VideoDecoder and Dav1dVideoDecoder?

[the remark this answers is not in the document]
Comment (victorpasoshnikov) [reply starboard/xb1/shared/gpu_base_video_decoder.cc]: > Ok, the new solution is simple enough. Just a last double check, can we confirm the drain operation are all finished synchronous in DecodeInternal() for VpxVideoDecoder, Av1VideoDecoder and Dav1dVideoDecoder?

Yes, we can. DecodeInternal(nullptr) causes all internal decoder's threads join. We can be sure that nothing happens in decoder until next DecodeInternal(not null) will be called.

[the remark this answers is not in the document]
Comment (TyHolc) [reply starboard/xb1/shared/gpu_base_video_decoder.cc]: @jasonzhangxx are there any other concerns or changes needed? I'm hoping to get this change into the next RC. And keep in mind that it has to go in tandem with [this change](https://lbshell-internal-review.googlesource.com/c/cobalt_src/+/267860)

**Review by amurovanyi — APPROVED** [review]

**Review by jasonzhangxx — APPROVED** [review]

**Pull request merged**